### PR TITLE
rgw: custom user data header

### PR DESCRIPTION
**Pull request by yehudasa**

### CODE REVIEW

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -2594,7 +2594,7 @@ class RGWPutObjProcessor_Multipart : public RGWPutObjProcessor_Atomic
   int do_complete(size_t accounted_size, const string& etag, real_time *mtime,
                   real_time set_mtime, map<string, bufferlist>& attrs,
                   real_time delete_at, const char *if_match,
-                  const char *if_nomatch) override;
+                  const char *if_nomatch, const string& user_data) override;
```
Comment (yehudasa) [inline src/rgw/rgw_op.cc]: @prallabh would rather see a string pointer here

File: src/rgw/rgw_op.cc [hunk 2]
```diff
@@ -3181,7 +3181,7 @@ void RGWPutObj::execute()
   }
 
   op_ret = processor->complete(s->obj_size, etag, &mtime, real_time(), attrs,
-                               (delete_at ? *delete_at : real_time()), if_match, if_nomatch);
+                               (delete_at ? *delete_at : real_time()), if_match, if_nomatch, user_data);
```
Comment (yehudasa) [inline src/rgw/rgw_op.cc]: `(user_data.empty() ? nullptr : &user_data)`

File: src/rgw/rgw_rados.cc [hunk 3]
```diff
@@ -2355,9 +2355,9 @@ void RGWObjVersionTracker::generate_new_write_ver(CephContext *cct)
 int RGWPutObjProcessor::complete(size_t accounted_size, const string& etag,
                                  real_time *mtime, real_time set_mtime,
                                  map<string, bufferlist>& attrs, real_time delete_at,
-                                 const char *if_match, const char *if_nomatch)
+                                 const char *if_match, const char *if_nomatch, const string& user_data)
```
Comment (yehudasa) [inline src/rgw/rgw_rados.cc]: string pointer

File: src/rgw/rgw_rados.cc [hunk 4]
```diff
@@ -2724,6 +2724,8 @@ int RGWPutObjProcessor_Atomic::do_complete(size_t accounted_size, const string&
   obj_op.meta.flags = PUT_OBJ_CREATE;
   obj_op.meta.olh_epoch = olh_epoch;
   obj_op.meta.delete_at = delete_at;
+  if (store->ctx()->_conf->rgw_swift_dlo_all_segments_size)
```
Comment (yehudasa) [inline src/rgw/rgw_rados.cc]: @prallabh why use this condition?

File: src/rgw/rgw_rados.cc [hunk 5]
```diff
@@ -9455,6 +9457,9 @@ int RGWRados::Bucket::UpdateIndex::complete(int64_t poolid, uint64_t epoch,
   ent.meta.accounted_size = accounted_size;
   ent.meta.mtime = ut;
   ent.meta.etag = etag;
+  if (store->ctx()->_conf->rgw_swift_dlo_all_segments_size && user_data)
```
Comment (yehudasa) [inline src/rgw/rgw_rados.cc]: this should only depend on user_data

File: src/rgw/rgw_rados.h [hunk 6]
```diff
@@ -2868,7 +2869,7 @@ class RGWRados
                    uint64_t accounted_size, ceph::real_time& ut,
                    const string& etag, const string& content_type,
                    bufferlist *acl_bl, RGWObjCategory category,
-		   list<rgw_obj_index_key> *remove_objs);
+		   list<rgw_obj_index_key> *remove_objs, const string *user_data = NULL);
```
Comment (yehudasa) [inline src/rgw/rgw_rados.h]: we should be using nullptr nowadays for null pointers, not NULL

File: src/rgw/rgw_rest_swift.cc [hunk 7]
```diff
@@ -306,6 +306,8 @@ void RGWListBucket_ObjStore_SWIFT::send_response()
       s->formatter->dump_string("name", key.name);
       s->formatter->dump_string("hash", iter->meta.etag);
       s->formatter->dump_int("bytes", iter->meta.accounted_size);
+      if (iter->meta.user_data.size() && s->cct->_conf->rgw_swift_dlo_all_segments_size)
+        s->formatter->dump_string("total_manifest_size", iter->meta.user_data);
```
Comment (yehudasa) [inline src/rgw/rgw_rest_swift.cc]: @prallabh don't call this field `total_manifest_size`. This is a custom user data field, and it's application dependent. Maybe `user_custom_data`.

File: src/rgw/rgw_rest_swift.cc [hunk 8]
```diff
@@ -758,6 +760,17 @@ int RGWPutObj_ObjStore_SWIFT::get_params()
     return r;
   }
 
+  if (s->cct->_conf->rgw_swift_dlo_all_segments_size) {
+    string custom_header = s->cct->_conf->rgw_swift_custom_header;
+    if (s->info.env->exists(custom_header.c_str())) {
+      user_data = s->info.env->get(custom_header.c_str());
+      if (!user_data.size()) {
```
Comment (yehudasa) [inline src/rgw/rgw_rest_swift.cc]: `if (user_data.empty())`

File: src/rgw/rgw_rest_swift.cc [hunk 9]
```diff
@@ -758,6 +760,17 @@ int RGWPutObj_ObjStore_SWIFT::get_params()
     return r;
   }
 
+  if (s->cct->_conf->rgw_swift_dlo_all_segments_size) {
```
Comment (yehudasa) [inline src/rgw/rgw_rest_swift.cc]: check `if (~s->cct->_conf->rgw_swift_custom_header.empty())` instead

File: src/rgw/rgw_rados.cc (hunk 4 above)
Comment (yehudasa) [inline src/rgw/rgw_rados.cc]: check `if (user_data)`

File: src/common/config_opts.h [hunk 10]
```diff
@@ -1679,3 +1679,7 @@ OPTION(event_tracing, OPT_BOOL, false) // true if LTTng-UST tracepoints should b
 OPTION(internal_safe_to_start_threads, OPT_BOOL, false)
 
 OPTION(debug_deliberately_leak_memory, OPT_BOOL, false)
+
+//This option is to return size of all segments for the manifest in DLO during bucket listing
+OPTION(rgw_swift_dlo_all_segments_size, OPT_BOOL, false)
```
Comment (yehudasa) [inline src/common/config_opts.h]: @prallabh drop this configurable

**Comment by prallabh** [comment]
@yehudasa thanks for the quick review, have addressed all the comments, please have a look.

File: src/common/config_opts.h [hunk 11]
```diff
@@ -1680,6 +1680,4 @@ OPTION(internal_safe_to_start_threads, OPT_BOOL, false)
 
 OPTION(debug_deliberately_leak_memory, OPT_BOOL, false)
 
-//This option is to return size of all segments for the manifest in DLO during bucket listing
-OPTION(rgw_swift_dlo_all_segments_size, OPT_BOOL, false)
-OPTION(rgw_swift_custom_header, OPT_STR, "")
+OPTION(rgw_swift_custom_header, OPT_STR, "") // option to enable swift custom headers
```
Comment (yehudasa) [inline src/common/config_opts.h]: @prallabh squash this change into the first commit, rewrite first commit title and message

File: src/rgw/rgw_rados.cc [hunk 12]
```diff
@@ -2724,6 +2724,8 @@ int RGWPutObjProcessor_Atomic::do_complete(size_t accounted_size, const string&
   obj_op.meta.flags = PUT_OBJ_CREATE;
   obj_op.meta.olh_epoch = olh_epoch;
   obj_op.meta.delete_at = delete_at;
+  if (user_data)
```
Comment (yehudasa) [inline src/rgw/rgw_rados.cc]: @prallabh no need for this if

**Comment by yehudasa** [comment]
@prallabh modify commit titles to have `rgw: ` prefix, remove all reference to dlo from commit messages

**Comment by prallabh** [comment]
@yehudasa squashed all of them into a single commit, please have a look, thanks!

**Comment by yehudasa** [comment]
@prallabh can you open a ceph tracker feature ticket on this one and add a reference to it in the commit message (with `Fixes:` tag). Otherwise looks good.

**Comment by prallabh** [comment]
@yehudasa updated with the tracker details

**Comment by prallabh** [comment]
@yehudasa have rebased, thanks.

**Comment by prallabh** [comment]
@yehudasa can you please help merge, thanks!

**Pull request merged**